src/privileged/notificationBar/api.js

```diff
@@ -0,0 +1,155 @@
+
+/* eslint-disable-next-line */
+var {EventManager, EventEmitter} = ExtensionCommon;
+/* eslint-disable-next-line no-var */
+var {Management: {global: {tabTracker}}} = ChromeUtils.import("resource://gre/modules/Extension.jsm", {});
+
+// eslint-disable-next-line no-undef
+XPCOMUtils.defineLazyModuleGetter(
+  this,
+  "BrowserWindowTracker",
+  "resource:///modules/BrowserWindowTracker.jsm",
+);
+
+/** Return most recent NON-PRIVATE browser window, so that we can
+ * manipulate chrome elements on it.
+ */
+function getMostRecentBrowserWindow() {
+  return BrowserWindowTracker.getTopWindow({
+    private: false,
+    allowPopups: false,
+  });
+}
+
+/** Display instrumented 'notification bar' explaining the feature to the user
+ *
+ *   Telemetry Probes:
+ *
+ *   - {event: survey-shown}
+ *
+ *   - {event: page-broken}
+ *
+ *   - {event: page-not-broken}
+ *
+ *    Note:  Bar WILL NOT SHOW if the only window open is a private window.
+ *
+ *    Note:  Handling of 'x' is not implemented.  For more complete implementation:
+ *
+ *      https://github.com/gregglind/57-perception-shield-study/blob/680124a/addon/lib/Feature.jsm#L148-L152
+ *
+ */
+class NotificationBarEventEmitter extends EventEmitter {
+  emitShow(variationName) {
+    const self = this;
+    const recentWindow = getMostRecentBrowserWindow();
+    const browser = recentWindow.gBrowser.selectedBrowser;
+    const tabId = tabTracker.getBrowserTabId(browser);
+
+    const primaryAction =  {
+      disableHighlight: true,
+      label: "Yes (broken)",
+      accessKey: "f",
+      callback: () => {
+        self.emit("page-broken", tabId);
+      },
+    };
+
+    const secondaryActions =  [
+      {
+        label: "No (works)",
+        accessKey: "d",
+        callback: () => {
+          self.emit("page-not-broken", tabId);
+        },
+      },
+    ];
+
+    // option name is described as: "An optional string formatted to look bold and used in the
+    //                    notification description header text. Usually a host name or
+    //                    addon name."
+    // It is bold, but not used in a header, we're working with it anyway.
+    const options = {
+      hideClose: true,
+      persistent: true,
+      autofocus: true,
+      name: "Firefox Survey: ",
+      popupIconURL: "chrome://branding/content/icon64.png",
+    };
+    recentWindow.PopupNotifications.show(browser, "fast-block-notification", "<> Did you reload this page to resolve loading issues?", null, primaryAction, secondaryActions, options);
+  }
+}
+
+this.notificationBar = class extends ExtensionAPI {
+  /**
+   * Extension Shutdown
+   * Goes through each tab for each window and removes the notification, if it exists.
+   */
+  onShutdown(shutdownReason) {
+    for (const win of BrowserWindowTracker.orderedWindows) {
+      for (const browser of win.gBrowser.browsers) {
+        const notification = win.PopupNotifications.getNotification("fast-block-notification", browser);
```

Comment:
Instead of this, we can just keep references to the notifications ourselves, and destroy them, instead of iterating over every tab. Some users have 100s of tabs.

Reply:
I'm not sure why iterating over 100s of things (in the worst case) should be slow and yes, keeping references is extra work and if done wrong causes us to leak the world, so I'd prefer not to.